### ISSUE TITLE
Stub functions should only typecheck for None

### ISSUE THREAD

**Issue by Wilfred** [python/mypy#4066]
This passes in mypy 0.521:

```python
def foo(): # type: () -> int
    pass
```

`mypy --py2 --strict demo.py` does not show any errors. If a function body does not contain a `return` statement or a `yield` statement it should only have a type of `None` I believe.

**Comment by ilevkivskyi** [comment]
This is a special case in mypy. If a function have an empty body then it can have arbitrary return type. This is useful for testing, and is necessary for stubs. We have lots of tests that use this, so we can't disable this in non-stub files. Alternatively we can somehow detect whether we are running tests, and display an error otherwise.

**Comment by JukkaL** [comment]
Mypy should probably eventually enforce this outside stubs. One way would be to make this an option (which doesn't need to be shown in help, perhaps) and turn it on by default, but we'd turn it off for most tests. Detecting stubs is easy, so it won't be a problem. We don't need to actually detect tests if all test runners just explicitly toggle the option.

**Comment by ilevkivskyi** [comment]
@JukkaL 
> We don't need to actually detect tests if all test runners just explicitly toggle the option.

Yes, I was thinking about the same idea.

**Comment by ilevkivskyi** [comment]
See also https://github.com/python/mypy/issues/2350

**Comment by Wilfred** [comment]
Aha, I didn't realise that mypy treated stubs specially. My intent was to minimise my example, but I initially encountered this issue on a function that calls exit:

```python
import sys

def fatal_error(message):
    # type: (str) -> dict
    print "Oh no! {}".format(message)
    sys.exit(1)
```

I was expecting mypy to give a type error here.

**Comment by JelleZijlstra** [comment]
That's expected: that function basically returns NoReturn, and NoReturn is a subtype of every type.

**Comment by gvanrossum** [comment]
I propose to change the title of this issue to "Get rid of exception for functions with just pass as body" -- or something better (but I can't think of something better). The current title is too wide and in general incorrect.

**Comment by Wilfred** [comment]
I've updated the issue title.

**Comment by Wilfred** [comment]
@JelleZijlstra the NoReturn behaviour was surprising to me. Could I open an issue for a `--strict` option that disabled this behaviour, or is this working as intended?

**Comment by ilevkivskyi** [comment]
> or is this working as intended?

This is intended. it is OK for a function to declare a return type wider than actual (``NoReturn`` in this case).

**Comment by Wilfred** [comment]
@JelleZijlstra out of interest, what's the benefit of that? When is `NoReturn` being a subtype of everything useful?

**Comment by ilevkivskyi** [comment]
@Wilfred ``NoReturn`` is the bottom type, i.e. a subtype of every type, so I would put this vice versa, what's the benefit of complicating the type system by _not_ doing that?

**Comment by ilevkivskyi** [comment]
@Wilfred  To clarify, this was a rhetorical question, this discussion is off-topic for this issue, if you want we can continue it elsewhere.

**Comment by ilevkivskyi** [comment]
I think this is just a duplicate of #2350
